velox/connectors/hive/iceberg/IcebergDataSink.cpp

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/iceberg/IcebergDataSink.h"
+
+#include <utility>
+#include "velox/common/base/Counters.h"
+#include "velox/common/base/Fs.h"
+#include "velox/connectors/hive/HiveConfig.h"
+#include "velox/connectors/hive/HivePartitionFunction.h"
+#include "velox/connectors/hive/TableHandle.h"
+#include "velox/core/ITypedExpr.h"
+#include "velox/exec/OperatorUtils.h"
+
+namespace facebook::velox::connector::hive::iceberg {
+
+namespace {
+#define WRITER_NON_RECLAIMABLE_SECTION_GUARD(index)       \
+  memory::NonReclaimableSectionGuard nonReclaimableGuard( \
+      writerInfo_[(index)]->nonReclaimableSectionHolder.get())
+} // namespace
+
+IcebergDataSink::IcebergDataSink(
+    facebook::velox::RowTypePtr inputType,
+    std::shared_ptr<const IcebergInsertTableHandle> insertTableHandle,
+    const facebook::velox::connector::ConnectorQueryCtx* connectorQueryCtx,
+    facebook::velox::connector::CommitStrategy commitStrategy,
+    const std::shared_ptr<const HiveConfig>& hiveConfig)
+    : HiveDataSink(
+          std::move(inputType),
+          std::move(insertTableHandle),
+          connectorQueryCtx,
+          commitStrategy,
+          hiveConfig) {}
+
+void IcebergDataSink::appendData(facebook::velox::RowVectorPtr input) {
+  checkRunning();
+
+  // Write to unpartitioned table.
+  if (!isPartitioned()) {
+    const auto index = ensureWriter(HiveWriterId::unpartitionedId());
+    write(index, input);
+    return;
+  }
+
+  dataChannels_ = getDataChannels(partitionChannels_, inputType_->size());
+
+  // Compute partition and bucket numbers.
+  computePartitionAndBucketIds(input);
+
+  // Lazy load all the input columns.
+  for (column_index_t i = 0; i < input->childrenSize(); ++i) {
+    input->childAt(i)->loadedVector();
+  }
+
+  // All inputs belong to a single non-bucketed partition. The partition id
+  // must be zero.
+  if (!isBucketed() && partitionIdGenerator_->numPartitions() == 1) {
+    const auto index = ensureWriter(HiveWriterId{0});
+    write(index, input);
+    return;
+  }
+
+  splitInputRowsAndEnsureWriters();
+
+  partitionData_.reserve(writers_.size());
+
+  const auto numRows = partitionIds_.size();
+
+  for (auto row = 0; row < numRows; ++row) {
```

Comment:
It is very inefficient to repeat this loop here just to populate partitionData_. It would be better to override splitInputRowsAndEnsureWriters() and populate partitionData_ there. Then we don't need to override appendData() at all.